### PR TITLE
feat(cluster): support migrate slot range

### PR DESCRIPTION
**Pull request by PokIsemaine**
issues: https://github.com/apache/kvrocks/issues/2355

This draft PR demonstrates how to support migrating slot ranges.

# What I Did

1 Migration job - 1 slot range:

* I encapsulated a `SlotRange` structure and changed the migration-related class members from a single slot to a slot range.
* Reference: https://github.com/apache/kvrocks/issues/412. Slot migration includes the following phases: start migration, migrate existing data, migrate incremental data, and end migration. In each modified phase, the entire slot range must be completed before moving to the next phase.

# TODO
TODO represents the items I hope to discuss.

1. Support multiple slot ranges:
   * Current situation: Only one migration job can be performed at a time, and each migration job corresponds to a slot range.
   * Possible modification: Allow multiple migration jobs, migrating sequentially or in parallel.

2. Perform multiple migrations consecutively but do not immediately use `setslot` to update the topology, referring to the example in TestSlotRangeMigrate:
```go
	t.Run("MIGRATE - Repeat migration cases, but does not immediately update the topology via setslot", func(t *testing.T) {
		// Disjoint
		require.Equal(t, "OK", rdb0.Do(ctx, "clusterx", "migrate", "114-116", id1).Val())
		waitForMigrateSlotRangeState(t, rdb0, "114-116", SlotMigrationStateSuccess)
		require.Equal(t, "OK", rdb0.Do(ctx, "clusterx", "migrate", "117-118", id1).Val())
		waitForMigrateSlotRangeState(t, rdb0, "117-118", SlotMigrationStateSuccess)
		require.Equal(t, "OK", rdb0.Do(ctx, "clusterx", "migrate", "112-113", id1).Val())
		waitForMigrateSlotRangeState(t, rdb0, "112-113", SlotMigrationStateSuccess)

		errMsg := "Can't migrate slot which has been migrated"
		// TODO: Migrating 112-113, but 114-118 is covered and cannot be detected.
		// require.ErrorContains(t, rdb0.Do(ctx, "clusterx", "migrate", "114-116", id1).Err(), errMsg)
		// require.ErrorContains(t, rdb0.Do(ctx, "clusterx", "migrate", "117-118", id1).Err(), errMsg)

		// Intersection
		require.ErrorContains(t, rdb0.Do(ctx, "clusterx", "migrate", "112", id1).Err(), errMsg)
		require.ErrorContains(t, rdb0.Do(ctx, "clusterx", "migrate", "112-112", id1).Err(), errMsg)
		require.ErrorContains(t, rdb0.Do(ctx, "clusterx", "migrate", "113", id1).Err(), errMsg)
		require.ErrorContains(t, rdb0.Do(ctx, "clusterx", "migrate", "113-113", id1).Err(), errMsg)

		// Subset
		require.ErrorContains(t, rdb0.Do(ctx, "clusterx", "migrate", "112-113", id1).Err(), errMsg)
		require.ErrorContains(t, rdb0.Do(ctx, "clusterx", "migrate", "112-120", id1).Err(), errMsg)
		require.ErrorContains(t, rdb0.Do(ctx, "clusterx", "migrate", "110-112", id1).Err(), errMsg)
	})
```
This situation also seems to exist in the original single slot migration, and I am not sure if such operations are reasonable.
```go
migrate A => migrate B => setslot A&B
migrate A => migrate B => migrate A (expected to fail, but allowed to pass) => setslot A&B

// slotrange A-C
migrate A-B => migrate C => setslot A-C
migrate A-B => migrate C => migrate B (expected error, but allowed to pass) => setslot A-C
```

3. More precise migration failure slot range:
   The current implementation determines the entire slot range to fail and cleans it up, and the user later migrates the entire slot range again.
   Do we want to support a more precise failure range? For example, `[start_slot-fail_slot), [fail_slot, end]`. Users can check the status with commands such as `cluster info` and then re-migrate the failed slot range by themselves. (Personally, I think it's a bit cumbersome and error-prone)

# Miscellaneous
Other suggestions are welcome, such as more testing, code optimization, better user interaction, etc.

### CODE REVIEW

File: src/cluster/cluster_defs.h
Comment (git-hulk) [inline src/cluster/cluster_defs.h]:
```suggestion
};

```

File: src/cluster/cluster_defs.h
Comment (git-hulk) [inline src/cluster/cluster_defs.h]: start <= end && end < kClusterSlots can deduce start < kClusterSlots
```suggestion
    return start >= 0 && end >= 0 && start <= end && end < kClusterSlots;
```

File: src/cluster/cluster_defs.h
Comment (git-hulk) [inline src/cluster/cluster_defs.h]: Should be `Contains`

File: src/cluster/cluster_defs.h
Comment (git-hulk) [inline src/cluster/cluster_defs.h]: I'm wondering if `-1` will be better than the "empty" literal since it can be parsed back to an invalid slot.

File: src/cluster/cluster_defs.h
Comment (git-hulk) [inline src/cluster/cluster_defs.h]: HasOverlap?

**Review by git-hulk — COMMENTED** [review]
@PokIsemaine Thanks for your impressive contributions. I will take another pass while it's ready to be reviewed.

File: src/storage/redis_db.cc
Comment (caipengbo) [inline src/storage/redis_db.cc]:
```suggestion
  return storage_->DeleteRange(prefix, prefix_end);
```

File: src/cluster/cluster_defs.h
Comment (caipengbo) [inline src/cluster/cluster_defs.h]: The semantics of range is preferably `[start,end)`.

File: src/storage/redis_db.cc
Comment (caipengbo) [reply src/storage/redis_db.cc]: Remove the old `ClearKeysOfSlot()`

File: src/cluster/slot_import.cc
Comment (caipengbo) [inline src/cluster/slot_import.cc]:
```suggestion
            fmt::format("only one importing job is allowed, current importing: {}", import_slot_range_.String())};
```

File: src/cluster/slot_import.cc
Comment (caipengbo) [inline src/cluster/slot_import.cc]:
```suggestion
    return {Status::NotOK, fmt::format("mismatch slots, importing slots: {}, but got: {}", import_slot_range_.String(),
```

File: src/cluster/slot_import.cc
Comment (caipengbo) [inline src/cluster/slot_import.cc]:
```suggestion
    return {Status::NotOK, fmt::format("mismatch slots, importing slots: {}, but got: {}", import_slot_range_.String(),
```

File: src/cluster/cluster_defs.h
Comment (PokIsemaine) [reply src/cluster/cluster_defs.h]: I referenced the `SetSlotRanges` implementation, and it should be following `[start, end]`.
https://github.com/apache/kvrocks/blob/7ed512e835201abe3f68f3c795dabb13acdc6733/src/cluster/cluster.cc#L120
Should they be harmonized, especially since `migrate` and `setslot` tend to appear together when the user is using them.
If you change it to `[start, end)` then `SetSlotRanges` should be changed synchronously too, what do you think to do?

Comment (caipengbo) [reply src/cluster/cluster_defs.h]: OK, keep it that way.

File: src/cluster/slot_import.cc
Comment (PokIsemaine) [reply src/cluster/slot_import.cc]: Thanks for the advice. Actually, if you search for `".String()"`, there are many similar issues.
image

In the beginning, I tried changing "slot" to "slots", but encountered some difficulties. On one hand, when `start == end`, `String()` will return a single `slot` instead of `[start,end]`, which causes ambiguity with the term "slots" in the message. On the other hand, some expected values in the test cases also need to be changed from "slot" to "slots". So I temporarily abandoned these modifications.

```cpp
std::string String() const {
  if (!IsValid()) return "-1";
  if (start == end) return fmt::format("{}", start);  // <== HERE
  return fmt::format("{}-{}", start, end);
}
```

Regarding the ambiguity issue, we could probably change it to "slot(s)" instead of "slot" or "slots" to be as compatible as possible with both representations. If you think that's OK, I'll change them (including some cases in the test cases).

Comment (caipengbo) [reply src/cluster/slot_import.cc]: Thanks for the deep thought. If you think it is not trouble, better modify it :)

**Comment by PokIsemaine** [comment]
@git-hulk @caipengbo 

If it's convenient, I would like to discuss some matters related to the TODO list:

First, regarding item 2 on the TODO list, I find this situation a bit confusing because I am not sure if this is the intended design of kvrocks. If possible, please determine whether this is a normal situation.

Second, concerning items 1 and 3 on the TODO list, these are enhancements. Please evaluate whether these features are needed and if they should be included in this PR.

**Comment by caipengbo** [comment]
@PokIsemaine 
> First, regarding item 2 on the TODO list, I find this situation a bit confusing because I am not sure if this is the intended design of kvrocks. If possible, please determine whether this is a normal situation.

This is indeed the case at the moment. The migrated slots is not checked when the migration is performed, and I think they should be checked.

> Second, concerning items 1 and 3 on the TODO list, these are enhancements. Please evaluate whether these features are needed and if they should be included in this PR.

I think these are good improvements, but we can add them in the subsequent PR. This PR is enough.

**Comment by PokIsemaine** [comment]
@caipengbo Ok, thanks for the suggestion. So for the second TODO, do we try to fix it at this PR or create another separate issue to track it?

**Comment by caipengbo** [comment]
> for the second TODO, do we try to fix it at this PR or create another separate issue to track it?

Let's fix it in this PR. @PokIsemaine

**Comment by git-hulk** [comment]
@PokIsemaine I think this PR is ready to review now?

**Comment by PokIsemaine** [comment]
> @PokIsemaine I think this PR is ready to review now?

You can start the review now :)

The `clang-tidy` check in the CI doesn't seem to be working, so please try to re-run it. This works in my local and fork repository CI.

**Comment by PragmaTwice** [comment]
Could you fix these clang-tidy issues? Please make sure the code can be successfully compiled.
```
/home/runner/work/kvrocks/kvrocks/build/_deps/fmt-src/include/fmt/base.h:1619:3: error: static_assert failed due to requirement 'formattable_pointer' "Formatting of non-void pointers is disallowed." [clang-diagnostic-error]
Suppressed 21102 warnings (21069 in non-user code, 32 NOLINT, 1 with check filters).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
Found compiler error(s).
  static_assert(formattable_pointer,
  ^             ~~~~~~~~~~~~~~~~~~~
/home/runner/work/kvrocks/kvrocks/build/_deps/fmt-src/include/fmt/base.h:2002:20: note: in instantiation of function template specialization 'fmt::detail::make_arg<true, fmt::context, double (double) noexcept, 0>' requested here
  return {{detail::make_arg<NUM_ARGS <= detail::max_packed_args, Context>(
                   ^
/home/runner/work/kvrocks/kvrocks/build/_deps/fmt-src/include/fmt/format.h:4357:28: note: in instantiation of function template specialization 'fmt::make_format_args<fmt::context, double (double) noexcept, std::basic_string<char>, 2UL, 0UL, 223ULL, 0>' requested here
  return vformat(fmt, fmt::make_format_args(args...));
                           ^
/home/runner/work/kvrocks/kvrocks/src/cluster/slot_migrate.cc:394:33: note: in instantiation of function template specialization 'fmt::format<double (&)(double) noexcept, std::basic_string<char> &>' requested here
    return {Status::NotOK, fmt::format("failed to iterate keys of slot {}: {}", slot, err_str)};
                                ^
/home/runner/work/kvrocks/kvrocks/src/cluster/slot_migrate.cc:393:66: error: use of undeclared identifier 'slot' [clang-diagnostic-error]
    LOG(ERROR) << "[migrate] Failed to iterate keys of slot " << slot << ": " << err_str;
                                                                 ^
/home/runner/work/kvrocks/kvrocks/src/cluster/slot_migrate.cc:394:81: error: use of undeclared identifier 'slot' [clang-diagnostic-error]
    return {Status::NotOK, fmt::format("failed to iterate keys of slot {}: {}", slot, err_str)};
                                                                                ^
```

**Comment by git-hulk** [comment]
> > @PokIsemaine I think this PR is ready to review now?
> 
> You can start the review now :)
> 
> The `clang-tidy` check in the CI doesn't seem to be working, so please try to re-run it. This works in my local and fork repository CI.

Thank you! I will take another pass in one or two days.

File: src/cluster/slot_migrate.cc
Comment (git-hulk) [inline src/cluster/slot_migrate.cc]: no_slot_range => empty_slot_range

File: src/cluster/slot_migrate.cc
Comment (git-hulk) [inline src/cluster/slot_migrate.cc:354]: Seems this assignment can be removed.

**Review by git-hulk — COMMENTED** [review]
This PR generally looks good to me, left two comments. To see if other maintainers have further comments.

File: src/cluster/slot_migrate.cc
Comment (PokIsemaine) [reply src/cluster/slot_migrate.cc:354]: This also informs the following error handling, considering the case where the first iter fails and does not update the `current_slot`.

```
  if (auto s = iter->status(); !s.ok()) {
    auto err_str = s.ToString();
    LOG(ERROR) << "[migrate] Failed to iterate keys of slot " << current_slot << ": " << err_str;
    return {Status::NotOK, fmt::format("failed to iterate keys of slot {}: {}", current_slot, err_str)};
  }
```

Comment (git-hulk) [reply src/cluster/slot_migrate.cc:354]: @PokIsemaine Got it, makes sense. Thanks for your explain.

**Review by git-hulk — DISMISSED** [review]

File: src/cluster/cluster_defs.h
Comment (mapleFU) [inline src/cluster/cluster_defs.h:48]: Perhaps a comment for inclusive is better here

File: src/cluster/cluster.cc
Comment (mapleFU) [inline src/cluster/cluster.cc]: Nit: a helper function for single range is ok here

File: src/storage/batch_extractor.cc
Comment (mapleFU) [inline src/storage/batch_extractor.cc:285]: Is `Contains` enough here?

File: src/storage/batch_extractor.cc
Comment (mapleFU) [inline src/storage/batch_extractor.cc:294]: ditto?

**Review by mapleFU — DISMISSED** [review]
+1 for we with some nits

File: src/storage/batch_extractor.cc
Comment (PokIsemaine) [reply src/storage/batch_extractor.cc:285]: Do you think there are any inappropriate examples? Could you please elaborate?

Comment (mapleFU) [reply src/storage/batch_extractor.cc:285]: Oh I got this since invalid always return !contains, ignoring this

File: src/cluster/cluster.cc
Comment (PokIsemaine) [reply src/cluster/cluster.cc]: Do you mean to keep `ClearKeysOfSlot`?

File: src/cluster/slot_migrate.h
Comment (mapleFU) [inline src/cluster/slot_migrate.h:197]: should we add a `static_assert(slot_range_.is_lock_free())` to force `slot_range` cas not uses mutex?

File: src/cluster/slot_migrate.h
Comment (mapleFU) [inline src/cluster/slot_migrate.h]: this is a bit weird so maybe rename the argument to `slow_range_in`?

Comment (PokIsemaine) [reply src/cluster/slot_migrate.h]: I don't quite understand the meaning of `slow_range_in`.

File: src/cluster/cluster.cc
Comment (mapleFU) [reply src/cluster/cluster.cc]: Nope, I mean maybe two helper functions:

```
// Get [slot_id, slot_id] 
SlotRange::GetPoint(int slot_id)
SlotRange::GetInvalid();
```

I think they would better for understanding

File: src/cluster/slot_migrate.h
Comment (PokIsemaine) [reply src/cluster/slot_migrate.h:197]: It seems like this can't be checked at compile time? `is_always_lock_free`?
https://godbolt.org/z/8qeW4M6he

File: src/cluster/slot_migrate.h
Comment (mapleFU) [reply src/cluster/slot_migrate.h]: `slot_range(slot_range)` is a bit weird since input arg and member shares the same naming. I think rename input argument to another name is a good way for this problem

File: src/cluster/slot_migrate.h
Comment (mapleFU) [reply src/cluster/slot_migrate.h:197]: Yeah, I think a compile time `is_always_lock_free` is better here, currently two integer will not meet any problem but If future somebody change this to more than 8B this should be take care of

File: src/cluster/slot_migrate.h
Comment (PokIsemaine) [reply src/cluster/slot_migrate.h]: ok, does `slow_range_in` here mean there is `max_speed`? I wanted to confirm if you actually meant to express `slot_range_in` but made a typo. Because I didn’t react right away.

File: src/cluster/cluster.cc
Comment (PokIsemaine) [reply src/cluster/cluster.cc]: I will add a static function `GetPoint` in `SlotRange`. For `GetInvaild`, I feel that it may not be suitable if it is used to replace `{-1, -1}`, because sometimes `{-1, -1}` represents a wrong SlotRange, and sometimes represents an empty SlotRange, depending on the context.

**Review by git-hulk — APPROVED** [review]

**Comment by git-hulk** [comment]
Will merge this PR if no further comments and CI is passed.

**Review by caipengbo — COMMENTED** [review]
LGTM

**Comment by mapleFU** [comment]
Nice work, thanks!

**Pull request merged**
